Add basic React 18 support

tkajtoch opened the pull request:
## Summary

This is the first set of changes needed to make EUI compatible with React 18. Following @cee-chen suggestion in #6772, I divided the work into more branches that are easier to review and merge into our feature branch (`feature/react-18`).

This PR contains:

- Upgrade react, react-dom and its types to v18
- Update package.json resolutions to force @types/react version to v18
- Fix component types to follow React 18 changes, most notably add `PropsWithChildren` to components accepting the `children` property
- Switch to using the new `createRoot` API in docs entrypoint

## QA

1. Pull the changes and run `yarn` to fetch new dependencies
2. Run `yarn build` and confirm the application builds successfully
3. Run `yarn start`, wait for it to load and open `http://localhost:8030` to confirm the application is loading
4. See the homepage and other component pages opening as expected

Please note that the work is not final and there are additional changes that need to be made to make the application stable and tests passing. At this stage, some pages may crash the application, portals are not working and the majority of tests are failing. This is expected and will be addressed in next PRs.

### General checklist

- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**

tkajtoch commented on src-docs/src/store/configure_store.js:
```diff
@@ -1,32 +1,33 @@
 export default function configureStore(initialState) {
   function rootReducer(state = {}, action) {
     return {
-      routing: routerReducer(state.routing, action),
```
I removed this because it wasn't working since upgrading from react-router v3 in June 2020 ([see commit](https://github.com/elastic/eui/commit/8d29c3115db72529a9a5d206328a6c027556b4d2)) and it doesn't seem like it's used anymore anyway. Keeping this would require us to switch to using `connected-react-router` instead.

cee-chen replied:
Do you mind quickly ELI5ing what this router reducer was even supposed to be doing? Quick smoke check - if our dark/light mode, babelfish switchers, and custom full screen routes (e.g. https://elastic.github.io/eui/#/templates/page-template/examples/centered-body) are working as expected, then awesome.

tkajtoch replied:
routerReducer updated the store whenever history location changed. It was a common practice till ~2019 to do so, to have an up-to-date location data straight in the store to use in redux middlewares, sagas, etc. 
It's not used anywhere anymore and the code didn't even work for the past 3 years because the library is not compatible with react-router v4 or newer, so I'm pretty confident this removal won't break anything :D

cee-chen commented on src-docs/src/views/tables/custom/custom.tsx:
```diff
@@ -291,12 +291,14 @@ export default class extends Component<{}, State> {
-            style={{ verticalAlign: 'text-top' }}
-          />{' '}
-          {title}
+          <>
```
That's odd, why was this extra fragment necessary? 😕

tkajtoch replied:
I added it first to supress the TS error `This JSX tag's 'children' prop expects a single child of type 'ReactNode', but multiple children were provided.` I thought is coming from the fact that span can't have block elements and it's related to the strict mode runtime checks in React 18, but as I just figured out it's happening because `title` can be either a `ReactNode` or `DataTitle`. We'll need to detect which one is provided and conditionally render `title` or `title.value`, but that's something for another PR to fix.

I removed it from this PR to not cause more confusion.

cee-chen commented on src-docs/src/views/tables/custom/custom.tsx:
```diff
@@ -671,6 +673,7 @@ export default class extends Component<{}, State> {
             align={column.alignment}
             truncateText={(cell as DataTitle)?.truncateText}
             textOnly={column.textOnly || false}
+            // @ts-ignore types of ReactNode and (item) => ReactNode are incompatible
```
It's a red flag if we need to `@ts-ignore` our own usage - we should either TODO this or figure out how to correctly type/cast our props correctly

tkajtoch replied:
I left it for further PRs to fix but I can add a TODO comment now if you'd like :)

tkajtoch replied:
Added a simple TODO comment to keep track of this

cee-chen reviewed (commented):
Hey Tomasz, huge apologies for taking so long to get to this PR while you were away. We had a fairly large set of dependency changes during your PTO (Node 18 upgrade being the biggest one that would affect this work). Would you mind rebasing/merging in latest main again, as well as running `yarn/yarn lint-fix`? (I'm seeing some newline/whitespace changes that are a little suspicious / that I wouldn't expect to see on latest Prettier).

Also I hate to be annoying, but I think we may need to either split up the type changes into another PR or go through each more slowly/carefully. There's way more changes than I would have expected to types (I was hoping it would be just the `children` change, but alas) and I'm relatively concerned about it. For better or for worse, we get a decent amount of Typescript questions and feedback, particularly around our components with generics, e.g. tables, select components, etc. - I'm a little surprised to see some of these types that used to be inferred now needing to be explicitly stated.

tkajtoch commented on src-docs/src/index.js:
```diff
@@ -40,98 +41,81 @@ const routes = [
+const root = createRoot(document.getElementById('guide'));
+
+root.render(
+  <StrictMode>
```
I'd say so. `StrictMode` runs additional checks during runtime and lets us know about potential issues with our code. Since the docs site is heavily used as a local development preview, I find it extremely handy to have it running React in strict mode.

I think the reason it's not enabled by default in React 18 is more for compatibility purposes than to let people choose whether they actually want their application to be strictly checked when running in development environments.

cee-chen replied:
I think my hesitation is purely around debugging. I can definitely see us "forgetting" or not realizing this is turned on locally, and thinking we can repro a bug in dev but not prod or vice versa. Personally, I think the most helpful option for us as frontend devs would be to add a toggle to our main header that allows us to toggle strict mode on or off - this will give us a visual hint that strict mode is enabled, and allow us to disable it dynamically to repro specific environments for specific bugs if necessary.

tkajtoch replied:
I'm not sure if there can be issues with non-strict mode when writing and testing code in strict mode. StrictMode also only affects development environments and EUI has it enabled only for the docs site. The main difference is that components are mounted twice when running in development mode to detect errors:

> To help surface these issues, React 18 introduces a new development-only check to Strict Mode. This new check will automatically unmount and remount every component, whenever a component mounts for the first time, restoring the previous state on the second mount.
([source](https://react.dev/blog/2022/03/08/react-18-upgrade-guide#updates-to-strict-mode))

I've never seen anyone having an option like that but maybe there's a reason for it. What do you say we leave it for now and add something like this if we actually find a non-StrictMode error we need to fix?

cee-chen replied:
Gotcha. Honestly it'll be a bit of an adjustment for me as well, I've never really worked in strict mode before and I wouldn't have expected, for example, double console logs due to remounting.

I might play around a bit later with adding a switcher for strict mode that goes next to our babelfish switch if you don't object - but that can go into the feature branch later.

tkajtoch commented on i18ntokens.json:
Sure! It made sense for me to commit it since it has quite a lot of changes due to the updates I did, but yeah if it's updated during releases anyway it doesn't really matter

tkajtoch commented on .babelrc.js:
```diff
@@ -11,8 +11,8 @@ module.exports = {
     }],
-    ["@babel/typescript", { isTSX: true, allExtensions: true }],
-    "@babel/react",
+    ["@babel/react", { runtime: 'classic' }],
+    ["@babel/typescript", { isTSX: true, allExtensions: true, allowDeclareFields: true }],
```
1. Order does matter for babel presets and plugins. In this case we want to transform JSX first so we can configure the transformation behavior. 
2. `runtime: 'classic'` enforces classic react runtime instead of trying to automatically detect what runtime you're using. The classic runtime expects `import React from 'react'`, the automatic runtime automatically imports it for you, but it didn't work flawlessly when running enzyme.

cee-chen replied:
Fantastic, thank you for the explanation!

tkajtoch commented on package.json:
```diff
@@ -220,17 +224,17 @@
-    "react-router": "^5.2.0",
-    "react-router-dom": "^5.2.0",
+    "react-router": "^5",
+    "react-router-dom": "^5",
```
Updated 👍🏻

tkajtoch commented on src/components/breadcrumbs/breadcrumb.tsx:
```diff
@@ -64,7 +65,8 @@ type _EuiBreadcrumbProps = {
   highlightLastBreadcrumb?: boolean;
   truncateLastBreadcrumb?: boolean;
-} & Pick<EuiBreadcrumbProps, 'truncate'>;
+} & Pick<EuiBreadcrumbProps, 'truncate'> &
+  PropsWithChildren;
 
 export const EuiBreadcrumb: FunctionComponent<
   HTMLAttributes<HTMLLIElement> & _EuiBreadcrumbProps
```
I actually wanted to discuss that with all of you as well. I think that we should include `PropsWithChildren` with the props type we define, so there's no difference between actual component props and props you import.

`{} & PropsWithChildren` / `interface Example extends PropsWithChildren` vs `PropsWithChildren<{}>` is another topic and I agree we should standardize it across our code base.

My vote goes to intersection (`&`) and extension (`extends`). It feels cleaner to me, can be nicely formatted, and doesn't use the generics syntax that could be confusing for some, especially that `PropsWithChildren` isn't really doing any generics tricks with the type it infers.

@cee-chen @1Copenut @breehall what do you think?

cee-chen replied:
++ to `&/extends` syntax over generic `<>` syntax! 👍

cee-chen replied:
@tkajtoch We have a majority on using the non-generic syntax it looks like - can you update all instances in this PR to use the agreed-upon syntax? Or alternatively, would you be open to separating out the types update to a separate PR to make reviewing a little easier?

tkajtoch commented on src-docs/src/index.js:
```diff
@@ -40,98 +41,81 @@ const routes = [
+                     isNew,
+                     component,
+                     from,
+                     to
```
Fixed! I think these lines got formatted automatically when resolving conflicts using my default IDE config.

cee-chen commented on i18ntokens.json:
This file is still present and needs to be configured to not have a diff / any changes from main

tkajtoch replied:
I removed the commits that update i18ntokens

cee-chen commented on yarn.lock:
The main feature branch and this PR still needs to be rebased/merged against latest main. I'm going to go ahead and push latest to the feature branch - LMK if you want any assistance untangling the yarn lock

tkajtoch commented:
@cee-chen I rebased the branch and removed most of the unrelated changes to keep reviewing as straightforward as possible. I have them stashed to include in separate PRs when this one is merged into our feature branch.

tkajtoch commented on src/components/portal/portal.tsx:
```diff
@@ -95,8 +95,8 @@ export class EuiPortal extends Component<EuiPortalProps> {
-      ? createPortal(this.props.children, this.portalNode)
-      : null;
+    return this.portalNode ? (
+      <>{createPortal(this.props.children, this.portalNode)}</>
```
It's wrapped in JSX fragment because createPortal returns `ReactPortal` which isn't a supported `FunctionComponent` return type

cee-chen replied:
Interesting - so the change is purely for typing and not because it affects actual end behavior? 🤔 

Just curious, why doesn't this affect these other instances in that case?

https://github.com/elastic/eui/blob/976e603bf55a41e6afe665488aa812e57a3bd47c/src/components/datagrid/utils/in_memory.tsx#L179

https://github.com/elastic/eui/blob/d549ff0d95969c977f299fbe63a732a0330f83d6/src/components/page_template/bottom_bar/page_bottom_bar.tsx#L71-L73

Or do you mean that class components no longer support this return type (since this component is a class, not function component)?

tkajtoch replied:
Yeah, sorry, I meant class components and `render()` return value!

cee-chen replied:
Gotcha!! Thanks for clarifying!

cee-chen commented on package.json:
```diff
@@ -147,8 +148,8 @@
     "@types/node": "^10.17.5",
-    "@types/react": "^17.0.38",
-    "@types/react-dom": "^17.0.11",
+    "@types/react": "^18.2.6",
```
It looks like this package has a few new patches we may want to grab
```suggestion
    "@types/react": "^18.2.14",
```

tkajtoch replied:
Bumped versions of all 4 of these!

cee-chen commented on package.json:
```diff
@@ -147,8 +148,8 @@
-    "@types/react": "^17.0.38",
-    "@types/react-dom": "^17.0.11",
+    "@types/react": "^18.2.6",
+    "@types/react-dom": "^18.2.4",
```
Same for this one
```suggestion
    "@types/react-dom": "^18.2.6",
```

tkajtoch replied:
Bumped versions of all 4 of these!

cee-chen commented on package.json:
```diff
@@ -56,7 +56,8 @@
   "resolutions": {
     "**/prismjs": "1.27.0",
-    "**/react": "^17.0.0"
+    "**/react": "^18.2.0",
+    "**/@types/react": "^18.2.6"
```
🤔 Do we need to specify the exact minors for these or can simply pin to a major?
```suggestion
    "**/react": "^18.0.0",
    "**/@types/react": "^18.0.0"
```

tkajtoch replied:
Yeah that's a good point. `^` would still do its job without the need to updating these lines when we upgrade to a newer minor version but it's definitely cleaner to just specify major. I replaced both with `^18` 👍🏻

cee-chen commented on src/components/code/code_block_annotations.tsx:
```diff
@@ -19,9 +24,10 @@ import { euiCodeBlockAnnotationsStyles } from './code_block_annotations.style';
-type EuiCodeBlockAnnotationProps = CommonProps & {
-  lineNumber: number;
-};
+type EuiCodeBlockAnnotationProps = PropsWithChildren &
+  CommonProps & {
```
Extra optional thought - if we're already exporting our own `CommonProps` type, would it make sense to just change it to `CommonPropsWithChildren` that extends `PropsWithChildren`? That way we only need one extra import? 🤷 

Probably not a huge deal in the long run, just throwing that thought out there

cee-chen replied:
No worries at all! Yeah it'd be a super minor syntactical sugar, just wanted to get your quick thoughts on whether you thought it was worth it. If you don't think it is, that's totally cool!

cee-chen commented on src/components/datagrid/body/data_grid_body_virtualized.tsx:
```diff
@@ -68,9 +69,15 @@ export const DataGridWrapperRowsContext =
     footerRow: null,
   });
 
+type InnerElementProps = PropsWithChildren & {
```
🙇 Thanks for cleaning this up!

cee-chen commented on src/components/form/form.styles.test.tsx:
```diff
@@ -17,7 +17,7 @@ import {
 } from './form.styles';
 
-const darkModeWrapper: React.FC = ({ children }) => (
+const darkModeWrapper: React.FC<PropsWithChildren> = ({ children }) => (
```
[super optional] Since we're already adding extra imports on line 9, we could as well grab `FunctionComponent` as well for consistency (this is on me btw, I wrote `React.FC` out of laziness when writing these tests originally 🙈)

```suggestion
const darkModeWrapper: FunctionComponent<PropsWithChildren> = ({ children }) => (
```

tkajtoch replied:
Updated 👍🏻

cee-chen commented on src/components/image/image_types.ts:
```diff
@@ -96,21 +101,23 @@ export type EuiImageWrapperProps = Pick<
-  isFullWidth: boolean;
-  setIsFullScreen: (isFullScreen: boolean) => void;
-};
+> &
+  PropsWithChildren & {
```
Also another annoying question that you are totally free to say you'd prefer not to enforce, but I wonder if we want to be more consistent with the order in which we declare `PropsWithChildren`? Would it be easier to tell if a component extends it at a glance if we declared it (and `CommonProps`) first? e.g.,

```ts
// Least to most specific
type EuiComponentProps = PropsWithChildren & CommonProps &
ExtendedComponentProps & {
  // ... props specific to this component
}
```

tkajtoch replied:
I love this idea! I updated the lines I changed to reflect that but I'd love us to document it actually make it a rule, just like module imports order for example.

cee-chen commented on src-docs/src/index.js:
```diff
@@ -40,130 +40,133 @@ const routes = [
   ...childRoutes,
 ];
 
-ReactDOM.render(
```
Random GitHub pro tip - using the `Hide whitespace` setting or adding `?w=0` to the PR URL is super helpful for added wrappers like these!

<img width="291" alt="" src="https://github.com/elastic/eui/assets/549407/60300424-400b-4962-a0bd-7478a188922a">

cee-chen commented:
Those are all my remaining comments - a lot of them are more general questions rather than change requests. This was a breeze to review, thanks for pruning it down Tomasz! ❤️‍🔥

cee-chen reviewed (approved):
Thanks for taking the time to answer my q's / address my comments, it's super appreciated!! I think I have one outstanding question I just wanted to get your thoughts on, but it's not a change request - feel free to merge in! 🚢 🎉
